File: 46-permutations/permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();

        backtracking(new ArrayList<>(), res, nums);

        return res;
    }

    public void backtracking(List<Integer> curr, List<List<Integer>> res, int[] nums) {
        if (curr.size() == nums.length) {
            res.add(new ArrayList<>(curr));
            return;
        }

        for (int num : nums) {
            if (!curr.contains(num)) {
                curr.add(num);

                backtracking(curr, res, nums);

                curr.remove(curr.size() - 1);
            }

        }
    }
}

File: 46-permutations/permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function (nums) {
    const n = nums.length;
    const res = [];
    const used = new Array(nums.length).fill(false);

    const backtracking = (arr) => {
        if (arr.length === n) {
            res.push([...arr]);
            return;
        }
        for (let i = 0; i < n; i++) {
            if (used[i]) continue;
            arr.push(nums[i]);
            used[i] = true;
            backtracking(arr);
            arr.pop();
            used[i] = false;;
        }
    }

    backtracking([]);
    return res;
};